### DISCUSSION TITLE
Why pnpm uses Hard links instead of soft links?

### DISCUSSION THREAD

**Discussion by khobaib222** [pnpm/pnpm#6800] [Q&A]
Pnpm installs modules in a global store and creates hard links to those packages in individual projects so that same packages do not take up extra disk space.
Is there a specific reason why we are using hard links instead of soft links here? Both type of links do not take up extra space.

**Comment by khobaib222** [comment]
@gluxon

**Reply by gluxon** [reply]
Looks like @zkochan and @thynson beat me to the answer. 🙂

If you're looking for a bit more detail, the Node.js docs highlight how it behaves when looking up modules through symlinks: https://nodejs.org/api/cli.html#--preserve-symlinks

> By default, when Node.js loads a module from a path that is symbolically linked to a different on-disk location, Node.js will dereference the link and use the actual on-disk "real path" of the module as both an identifier and as a root path to locate other dependency modules. In most cases, this default behavior is acceptable. However, **when using symbolically linked peer dependencies, as illustrated in the example below, the default behavior causes an exception to be thrown** if moduleA attempts to require moduleB as a peer dependency:

Thus, pnpm uses hard links instead of symlinks where appropriate to prevent scenarios like the above.

**Comment by thynson** [comment]
The most important reason is that node treats hard links as different files, just like two or more files with same content.

**Comment by zkochan** [comment]
Node.js uses the file's real location to resolve dependencies.

Let's say we have two deps. Dep and dep-dep. dep-dep is required from dep.

```
node_modules
  dep
    index.js ---> ~/.pnpm-store/files/00/a92134[..]023
  dep-dep
```

If we will use symlinks Node.js will try to resolve `require('dep-dep')` from `~/.pnpm-store/files/00/a92134[..]023`, so it won't be able to resolve it.
